### PR TITLE
cartesian_config: can't replace variable with names with "-"

### PR DESCRIPTION
**Pull request by bgartzi**
Value substitutions of variables with a "-" are now parsed properly.
The scope of the ignored exception has been minimized to KeyError so other possible exceptions are not masked.

ID: 2027950

### CODE REVIEW

**Comment by bgartzi** [comment]
@YongxueHong @luckyh could you have a check on this pull request please?

File: virttest/cartesian_config.py [hunk 1]
```diff
@@ -606,15 +606,15 @@ def _substitution(value, d):
     if "$" in value:
         start = 0
         st = ""
+        d_flat = _drop_suffixes(d)
```
Comment (luckyh) [inline virttest/cartesian_config.py:609]: Good catch, just suggest splitting this into another commit, thanks.

File: selftests/unit/test_cartesian_config.py [hunk 2]
```diff
@@ -387,9 +387,11 @@ def testVariableAssignment(self):
                     var += a
                     var <= b
                     system = 2
-                    ddd = ${tests + str(int(system) + 3)}4
-                    error = ${tests + str(system + 3)}4
-                    s.* ?= ${tests + "ahoj"}4
+                    variable-name-with-dashes = sampletext
+                    ddd = tests variant is ${tests}
+                    dashes = show ${variable-name-with-dashes}
```
Comment (luckyh) [inline selftests/unit/test_cartesian_config.py:392]: Additionally, I'd suggest adding the following tests as well.
```
1st = 1
starts_with_number = index ${1st}

not_a_substitution = ${}
```

**Comment by luckyh** [comment]
@ldoktor @pevogam @kylazhang @chunfuwen your review is appreciated.

**Review by luckyh — APPROVED** [review]
LGTM

**Comment by pevogam** [comment]
Great, I will run some testing first as we use lots of heavy Cartesian configuration.

**Review by YongxueHong — APPROVED** [review]

**Review by pevogam — CHANGES_REQUESTED** [review]
Hmmm this breaks some default behaviors for us:
```
[stdlog] 2021-12-20 17:25:12,673 test             L0310 INFO | INIT 1-CmdParserTest.test_selected_vms
[stdlog] 2021-12-20 17:25:12,675 parameters       L0142 DEBUG| PARAMS (key=timeout, path=*, default=None) => None
[stdlog] 2021-12-20 17:25:12,675 test             L0338 DEBUG| Test metadata:
[stdlog] 2021-12-20 17:25:12,675 test             L0340 DEBUG|   filename: /root/rpmbuild/QA/BUILD/avocado-i2n-92.0/selftests/test_cmd_parser.py
[stdlog] 2021-12-20 17:25:12,676 test             L0346 DEBUG|   teststmpdir: /var/tmp/avocado_gdds8wu0
[stdlog] 2021-12-20 17:25:12,676 test             L0536 INFO | START 1-CmdParserTest.test_selected_vms
[stdlog] 2021-12-20 17:25:12,677 test             L0207 DEBUG| DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
[stdlog] 2021-12-20 17:25:12,698 cmd_parser       L0106 DEBUG| Parsed param dict '{'aaa': 'bbb'}'
[stdlog] 2021-12-20 17:25:12,719 cmd_parser       L0174 DEBUG| Parsed vm strings 'OrderedDict([('vm1', ''), ('vm2', ''), ('vm3', '')])'
[stdlog] 2021-12-20 17:25:12,725 cmd_parser       L0203 DEBUG| Parsed tests string 'only normal
[stdlog] '
[stdlog] 2021-12-20 17:25:12,752 stacktrace       L0039 ERROR| 
[stdlog] 2021-12-20 17:25:12,752 stacktrace       L0041 ERROR| Reproduced traceback from: /usr/lib/python3.9/site-packages/avocado/core/test.py:793
[stdlog] 2021-12-20 17:25:12,753 stacktrace       L0045 ERROR| Traceback (most recent call last):
[stdlog] 2021-12-20 17:25:12,753 stacktrace       L0045 ERROR|   File "/root/rpmbuild/QA/BUILD/avocado-i2n-92.0/selftests/test_cmd_parser.py", line 33, in test_selected_vms
[stdlog] 2021-12-20 17:25:12,753 stacktrace       L0045 ERROR|     self.assertIn("only CentOS", self.config["vm_strs"]["vm1"])
[stdlog] 2021-12-20 17:25:12,753 stacktrace       L0045 ERROR|   File "/usr/lib64/python3.9/unittest/case.py", line 1098, in assertIn
[stdlog] 2021-12-20 17:25:12,754 stacktrace       L0045 ERROR|     self.fail(self._formatMessage(msg, standardMsg))
[stdlog] 2021-12-20 17:25:12,754 stacktrace       L0045 ERROR|   File "/usr/lib/python3.9/site-packages/avocado/core/test.py", line 979, in fail
[stdlog] 2021-12-20 17:25:12,754 stacktrace       L0045 ERROR|     raise exceptions.TestFail(message)
[stdlog] 2021-12-20 17:25:12,754 stacktrace       L0045 ERROR| avocado.core.exceptions.TestFail: 'only CentOS' not found in ''
```
I will have to investigate and report back.

File: selftests/unit/test_cartesian_config.py (hunk 2 above)
Comment (pevogam) [reply selftests/unit/test_cartesian_config.py:392]: This comment is not resolved but with the LGTM on @luckyh's side I understand it as such and won't repeat any feedback already provided.

File: virttest/cartesian_config.py (hunk 1 above)
Comment (pevogam) [reply virttest/cartesian_config.py:609]: Indeed, and this is performance critical location, great!

File: virttest/cartesian_config.py [hunk 3]
```diff
@@ -606,15 +606,15 @@ def _substitution(value, d):
...
             match = match_substitute.search(value, start)
             while match:
-                d_flat = _drop_suffixes(d)
-                val = eval(match.group(1), None, d_flat)
+                val = d_flat[match.group(1)]
                 st += value[start:match.start()] + str(val)
                 start = match.end()
                 match = match_substitute.search(value, start)
-        except Exception:
+        except KeyError:
```
Comment (pevogam) [inline virttest/cartesian_config.py:617]: I assume investigation in the original commit where this code was introduced does not mention why the choise used to be an `Exception` instance. That is, I hope this change will not introduce some issues some issues the author was trying to contain. My testing did not reveal problem here so I guess it is reasonable.

**Review by pevogam — APPROVED** [review]
I investigated and this PR wasn't the cause of it so LGTM. Reading through the commit messages I had to poll some additional opinions on whether it makes sense to revert the operations the reverted commit enabled. As it turned out nobody was really using and even aware of these additional operations, this PR good to go and focus the Cartesian config on simple string substitution.

**Comment by luckyh** [comment]
I'm going to merge this since we have no further comments against it for a month and there is already 3 approvals.

Thanks to all.

**Pull request merged**